src/components/NftPreview/NftPreview.tsx

```diff
@@ -45,32 +45,38 @@ function NftPreview({
           ...NftPreviewAssetFragment
         }
         collection @required(action: THROW) {
+          # TODO: what's the difference between id and dbid?
           id
           dbid
+          # TODO: can we make these required in the schema?
+          gallery @required(action: THROW) {
+            owner @required(action: THROW) {
+              username @required(action: THROW)
+            }
+          }
```

Comment:
first off, you can def mark them as required as it's true that collections will always have a gallery / owner / username

as for why the schema isn't labeled `Gallery!`:

![image](https://user-images.githubusercontent.com/12162433/178200465-c933cae9-e045-4620-9e3c-9a9b754c2a49.png)

i believe it's because it's marked as `@goField(forceResolver: true)` which is a way of denoting "we'll only resolve this data if you ask for it" bc it triggers a sub-query on the backend. whereas if it was required then the backend might always try to grab it.

@radazen can you confirm?

Reply:
(in the meantime ezra i shared the link to your academy award winning graphql/api presentation)

Reply:
I mentioned this to @Robinnnnn already but I think the thing that had me confused is that I am used to viewing the GraphQL schema as more of a guarantee on what data is stored in the backend rather than a guarantee on how the data is retrieved.

For example, it's unclear if a "gallery has an optional owner" means there are gaps data for some reason (technical debt, or a feature I'm unaware of). And inversely, if an optional field is truly meant to be optional, it's probably not a great idea for me to willy-nilly add `@required`. So I'm left wondering: what are our actual data relationships and what guarantees are there in those relationships (i.e. nullable or non-nullable).

Reply:
Here's an example where I'm at a loss around what to do:

<img width="1051" alt="image" src="https://user-images.githubusercontent.com/508855/178387977-976a31dc-5c14-482a-8d12-abe53aed4992.png">
<img width="783" alt="image" src="https://user-images.githubusercontent.com/508855/178387997-70e08d62-4824-456e-ab53-557f34f28b14.png">

I didn't modify/affect this code AFAIK, but getting a runtime error due to a query requiring a prop/data that appears to be missing? Makes it hard to know which things are safe to require or not.

Reply:
looping in @tbezman for the first half of the convo :P

re: erroring feed events it's bc the user who created those events were deleted on dev. will address separately

Reply:
@holic 

> For example, it's unclear if a "gallery has an optional owner" means there are gaps data for some reason (technical debt, or a feature I'm unaware of). And inversely, if an optional field is truly meant to be optional, it's probably not a great idea for me to willy-nilly add @required. So I'm left wondering: what are our actual data relationships and what guarantees are there in those relationships (i.e. nullable or non-nullable).

I try to think about "null" as a cop out for "optional". Any place in our schema where something might not be there for a "good reason", we use [union types in GraphQL](https://graphql.org/learn/schema/#union-types) to represent that. [Here's an example where we do that in our schema](https://github.com/gallery-so/gallery/blob/main/schema.graphql#L360).

At Gallery, `null` is used to represent "unhandled error", or an error that the client' can't do anything reasonable with. Something that would make the client throw their hands up and show a "Something went wrong" message.

The Relay team has an excellent write up [here](https://relay.dev/docs/guides/required-directive/#why-not-implement-this-at-the-schemaserver-level).

---

On a separate note, I would encourage you to try not to use `@required` if possible. Try and think "If the server can't get me this field, can I just not show the HTML for that field? Can I show a sane fallback instead?". 

`@required` is the least graceful method of handling a failure.

---

Honestly a lot of this stuff comes from dealing with the alternatives. At this point in time w/ GraphQL & Relay, this is the most sane way to handle this. I would be happy to get on a call with you and explain how we arrived at this conclusion.

Reply:
@Robinnnnn this will most likely continue to be a pain point for engineers working on the codebase. We should write something up explaining the philosophy to avoid these long conversations in the future.

Reply:
> I would be happy to get on a call with you and explain how we arrived at this conclusion.

Would love this! I've not worked within this paradigm and the context would be super helpful.

Reply:
@tbezman Just had a thought along this line, and correct me if I am wrong, but because we specify fragments for our components' data needs, we're no longer in control of when/where that query (i.e. a batch of fragments) gets run, therefore if we have `@required(action: THROW)`, we risk a whole ~page worth of data not returning due to one deep (and potentially inconsequential) field not resolving. Is that an accurate picture of why `@required(action: THROW)` is kind of a bad habit here?

Reply:
Yeah that's definitely a good point to bring  up. You may not know where the nearest Error Boundary is (relative to this component).

I think the easiest and most convincing question you can ask yourself is "Is there something better than an error here?"

I'll reach out on Discord to setup a time.